File: src/SmeAccounting.Web/Models/EnterpriseCode.cs
using System.Text.RegularExpressions;

namespace SmeAccounting.Web.Models;

public partial class EnterpriseCode : IEquatable<EnterpriseCode>
{
    private static readonly Regex FormatRegex = EnterpriseCodeRegex();

    public string Value { get; }

    private EnterpriseCode(string value) => Value = value;

    public static Result<EnterpriseCode> Create(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            return Result<EnterpriseCode>.Failure("Enterprise code format: must be 10 digits");

        if (!FormatRegex.IsMatch(value))
            return Result<EnterpriseCode>.Failure("Enterprise code format: must be exactly 10 digits (^\\d{10}$)");

        return Result<EnterpriseCode>.Success(new EnterpriseCode(value));
    }

    public bool Equals(EnterpriseCode? other) => other is not null && Value == other.Value;
    public override bool Equals(object? obj) => obj is EnterpriseCode other && Equals(other);
    public override int GetHashCode() => Value.GetHashCode();
    public override string ToString() => Value;
    public static implicit operator string(EnterpriseCode code) => code.Value;

    [GeneratedRegex(@"^\d{10}$")]
    private static partial Regex EnterpriseCodeRegex();
}

File: server/src/domain/valueObjects/EnterpriseCode.ts
const ENTERPRISE_CODE_PATTERN = /^\d{10}$/;

export class EnterpriseCode {
  private constructor(private readonly value: string) {}

  static create(raw: string): EnterpriseCode {
    const cleaned = raw.replace(/[.\-\s]/g, '');
    if (!ENTERPRISE_CODE_PATTERN.test(cleaned)) {
      throw new Error('Invalid enterprise code: must be exactly 10 digits');
    }
    return new EnterpriseCode(cleaned);
  }

  toString(): string {
    return this.value;
  }

  equals(other: EnterpriseCode): boolean {
    return this.value === other.value;
  }

  getRegistrationYear(): number {
    const yearPrefix = parseInt(this.value.substring(0, 2), 10);
    return 2000 + yearPrefix;
  }

  getProvinceCode(): string {
    return this.value.substring(2, 4);
  }

  getSerialNumber(): string {
    return this.value.substring(4);
  }
}
